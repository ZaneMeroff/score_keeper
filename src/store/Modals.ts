// Decorators
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

@Module
export default class Modals extends VuexModule {

  addPlayersModal: boolean = false
  clearDataModal: boolean = false
  deletePlayersModal: boolean = false
  resetScoreModal: boolean = false
  scoreLimitsModal: boolean = false

  get getShowAddPlayersModal(): boolean {
    return this.addPlayersModal
  }

  get getShowClearDataModal(): boolean {
    return this.clearDataModal
  }

  get getShowDeletePlayersModal(): boolean {
    return this.deletePlayersModal
  }

  get getShowResetScoreModal(): boolean {
    return this.resetScoreModal
  }

  get getShowScoreLimitsModal(): boolean {
    return this.scoreLimitsModal
  }

  // ------------------------------------------
  //       add players modal visibility
  // ------------------------------------------
  
  @Mutation
  addPlayersModalVisibility(show: boolean) {
    this.addPlayersModal = show
  }
  
  @Action
  action_addPlayersModalVisibility(show: boolean) {
    this.addPlayersModalVisibility(show)
  }

  // ------------------------------------------
  //        clear data modal visibility
  // ------------------------------------------
  
  @Mutation
  clearDataModalVisibility(show: boolean) {
    this.clearDataModal = show
  }
  
  @Action
  action_clearDataModalVisibility(show: boolean) {
    this.clearDataModalVisibility(show)
  }

  // ------------------------------------------
  //  confirm delete players modal visibility
  // ------------------------------------------
  
  @Mutation
  deletePlayersModalVisibility(show: boolean) {
    this.deletePlayersModal = show
  }

  @Action
  action_deletePlayersModalVisibility(show: boolean) {
    this.deletePlayersModalVisibility(show)
  }

  // ------------------------------------------
  //    confirm reset score modal visibility
  // ------------------------------------------
  
  @Mutation
  resetScoreModalVisibility(show: boolean) {
    this.resetScoreModal = show
  }

  @Action
  action_resetScoreModalVisibility(show: boolean) {
    this.resetScoreModalVisibility(show)
  }

  // ------------------------------------------
  //        score limit modal visibility
  // ------------------------------------------
  
  @Mutation
  scoreLimitModalVisibility(show: boolean) {
    this.scoreLimitsModal = show
  }

  @Action
  action_scoreLimitModalVisibility(show: boolean) {
    this.scoreLimitModalVisibility(show)
  }
}