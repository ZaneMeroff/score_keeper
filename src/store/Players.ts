// Vue & Decorators
import Vue from 'vue'
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

// Types
import { Player, PlayerData } from '@/types/players'

@Module
export default class Players extends VuexModule {

    playerData: PlayerData = {}

    get getPlayerData() {
      return this.playerData
    }

    // ------------------------------------------
    //              create player
    // ------------------------------------------

    @Mutation
    createPlayer(payload: Player) {
      Vue.set(this.playerData, payload.id, payload)
    }

    @Action
    action_createPlayer(payload: Player) {
      this.createPlayer(payload)
    }

    // ------------------------------------------
    //         modify target player score
    // ------------------------------------------

    @Mutation
    setPlayerScore(payload: { id: string, score: number }) {
      this.playerData[payload.id].score = payload.score
    }

    @Action
    action_setPlayerScore(payload: { id: string, score: number }) {
      this.setPlayerScore(payload)
    }

    // ------------------------------------------
    //         modify target player name
    // ------------------------------------------

    @Mutation
    setPlayerName(payload: { id: string, name: string }) {
      this.playerData[payload.id].name = payload.name
    }

    @Action
    action_setPlayerName(payload: { id: string, name: string }) {
      this.setPlayerName(payload)
    }

    // ------------------------------------------
    //        zero scores for all players
    // ------------------------------------------

    @Mutation
    zeroScores() {
      for (const player in this.playerData) {
        this.playerData[player].score = 0
      }
    }

    @Action
    action_zeroScores() {
      this.zeroScores()
    }
}