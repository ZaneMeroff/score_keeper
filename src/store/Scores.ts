// Decorators
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

// Types
import { Score, ScoreData } from '@/types/scores'
import Vue from 'vue'

@Module
export default class Scores extends VuexModule {

    scoreData: ScoreData = {}

    get getScoreData() {
      return this.scoreData
    }

    @Mutation
    setScoreData(payload: { id: string, updatedScore: number }) {
      // create or update score data based on id
      Vue.set(this.scoreData, payload.id, payload.updatedScore)
    }

    
}