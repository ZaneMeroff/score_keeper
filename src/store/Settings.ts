// Decorators
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

@Module
export default class Settings extends VuexModule {

  darkMode: boolean = true
  minScore: number = 0
  maxScore: number = 10
  numSelectorRangeSlides: string[] = []

  get getDarkMode(): boolean {
    return this.darkMode
  }

  get getDarkModeText(): 'color: #FFF' | 'color: #000' {
    return this.darkMode ? 'color: #FFF' : 'color: #000'
  }

  get getMinScore(): number {
    return this.minScore
  }

  get getMaxScore(): number {
    return this.maxScore
  }

  get getNumSelectorRangeSlides(): string[] {
    return this.numSelectorRangeSlides
  }
  
  // ------------------------------------------
  //             toggle dark mode
  // ------------------------------------------

  @Mutation
  setDarkMode() {
    this.darkMode = !this.darkMode
  }

  @Action
  action_setDarkMode() {
    this.setDarkMode()
  }

  // ------------------------------------------
  //            set score limits
  // ------------------------------------------

  @Mutation
  setScoreLimits(payload: { min: number, max: number }) {
    this.minScore = payload.min
    this.maxScore = payload.max
  }

  @Action
  action_setScoreLimits(payload: { min: number, max: number }) {
    this.setScoreLimits(payload)
    this.action_setNumSelectorRangeSlides()
  }

  // ------------------------------------------
  //      set score slides for NumSelector
  // ------------------------------------------
  
  @Mutation
  setNumSelectorRangeSlides() {
    // clear existing slides array
    this.numSelectorRangeSlides.splice(0)
    // create new slides based on min/max limits
    for (let i = this.minScore; i <= this.maxScore; i++) {
      this.numSelectorRangeSlides.push(i.toString())
    }
  }

  @Action
  action_setNumSelectorRangeSlides() {
    this.setNumSelectorRangeSlides()
  }
}