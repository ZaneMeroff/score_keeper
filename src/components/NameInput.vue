<template>

  <div>

    <q-item-section class="q-item-section">
      <div class="flex">
      <!-- edit btn -->
        <q-btn flat @click="() => showNameModal = true">
          <q-icon name="edit"></q-icon>
        </q-btn>
      <!-- player name -->
        <p class="player-name">{{ statePlayers.getPlayerData[playerId].name }}</p>
      </div>
      <!-- delete btn -->
      <q-btn flat @click="() => showDeleteModal = true">
        <q-icon name="delete"></q-icon>
      </q-btn>
    </q-item-section>

    <!-- edit player name modal -->
    <player-name 
      :showModal="showNameModal"
      :playerId="playerId"
      @close="() => showNameModal = false"
    />

    <!-- confirm delete modal -->
    <confirm 
      :showModal="showDeleteModal"
      :text="`Delete ${getName}?`"
      :onYes="() => {
        statePlayers.action_deletePlayer(playerId)
        showDeleteModal = false
      }"
      :onNo="() => showDeleteModal = false"
    />

  </div>

</template>

<script lang="ts">

// Decorators
import { Component, Prop, Vue } from 'vue-property-decorator';

// Vuex
import { statePlayers } from '@/store/index'

// Components
import Confirm from './modals/Confirm.vue'
import PlayerName from './modals/PlayerName.vue'

@Component({
  components: {
    'confirm': Confirm,
    'player-name': PlayerName
  }
})
export default class NameInput extends Vue {

  @Prop(String) private playerId!: string

  statePlayers = statePlayers
  
  showDeleteModal: boolean = false
  showNameModal: boolean = false

  get getName(): string {
    return this.statePlayers.getPlayerData[this.playerId].name
  }
}
</script>

<style scoped>

  .q-item-section {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    width: 100%;
  }

  .player-name {
    font-size: 16px;
    padding-top: 5px;
    margin: 0px;
  }

</style> 