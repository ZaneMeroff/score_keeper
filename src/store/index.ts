import Vue from 'vue'
import Vuex from 'vuex'
import VuexPersistence from 'vuex-persist'
import localforage from 'localforage'

import Modals from './Modals'
import Players from './Players'
import Settings from './Settings'

Vue.use(Vuex)

const vuexLocal = new VuexPersistence({
  storage: localforage as any,
  asyncStorage: true,
  modules: [ 'Modals', 'Players', 'Settings' ],
});

export const store: any = new Vuex.Store({
  // strict: process.env.DEV === 'true',
  state: {},
  modules: {},
  plugins: [ vuexLocal.plugin ],
})

export const stateModals= new Modals({ store, name: 'Modals' })
export const statePlayers= new Players({ store, name: 'Players' })
export const stateSettings = new Settings({ store, name: 'Settings' })
