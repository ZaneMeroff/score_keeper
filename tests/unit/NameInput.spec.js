import { createLocalVue, shallowMount } from '@vue/test-utils'
import { statePlayers } from '@/store/index'
import NameInput from '@/components/NameInput.vue'
import Quasar, * as All from 'quasar'
import uuid from 'uuid/v4'
import Vuex from 'vuex'

// --------- Vuex & Quasar setup -------------

const localVue = createLocalVue()
localVue.use(Vuex, Quasar, {components: All, directives: All, plugins: All})

// ---------------- mocks --------------------

jest.mock('uuid/v4')
uuid.mockImplementation(() => '12345')

// ------------ create player ----------------

statePlayers.action_createPlayers(1)

// -------------------------------------------

describe('NameInput', () => {

  describe('snapshots', () => {

    it('should render with default html', () => {
      const propsData = { playerId: '12345' }
      const component = shallowMount(NameInput, { localVue, propsData })
      
      expect(component.element).toMatchSnapshot()
    })
  })

  describe('props', () => {

    describe('playerId', () => {

      it('should accept a string', () => {
        const propsData = { playerId: '12345' }
        const component = shallowMount(NameInput, { localVue, propsData })
        
        expect(component.vm.$props.playerId).toEqual(propsData.playerId)
      })
    })
  })

  describe('data', () => {

    describe('showDeleteModal', () => {

      it('should have a default value of false', () => {
        const propsData = { playerId: '12345' }
        const component = shallowMount(NameInput, { localVue, propsData })
        
        expect(component.vm.$data.showDeleteModal).toEqual(false)
      })

      it('should accept a boolean', () => {
        const propsData = { playerId: '12345' }
        const component = shallowMount(NameInput, { localVue, propsData })
        const newValue = true

        component.setData({ showDeleteModal: newValue })

        expect(component.vm.$data.showDeleteModal).toEqual(newValue)
      })
    })

    describe('showNameModal', () => {

      it('should have a default value of false', () => {
         const propsData = { playerId: '12345' }
         const component = shallowMount(NameInput, { localVue, propsData })
         
         expect(component.vm.$data.showNameModal).toEqual(false)
       })
 
       it('should accept a boolean', () => {
         const propsData = { playerId: '12345' }
         const component = shallowMount(NameInput, { localVue, propsData })
         const newValue = true
 
         component.setData({ showNameModal: newValue })
 
         expect(component.vm.$data.showNameModal).toEqual(newValue)
       })
    })
  })

  describe('getters', () => {
  
    describe('getName', () => {
      
      it('should return correct name', () => {
        const propsData = { playerId: '12345' }
        const component = shallowMount(NameInput, { localVue, propsData })

        const expected = 'Player 1'
        expect(component.vm.getName).toEqual(expected)
      })
    })
  })
})
