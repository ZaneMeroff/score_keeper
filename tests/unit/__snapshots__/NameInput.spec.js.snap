// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameInput snapshots should render with default html 1`] = `
<div>
  <q-item-section-stub
    class="q-item-section"
  >
    <div
      class="flex"
    >
      <q-btn-stub
        align="center"
        flat="true"
        ripple="true"
      >
        <q-icon-stub
          name="edit"
          tag="i"
        />
      </q-btn-stub>
       
      <p
        class="player-name"
      >
        Player 1
      </p>
    </div>
     
    <q-btn-stub
      align="center"
      flat="true"
      ripple="true"
    >
      <q-icon-stub
        name="delete"
        tag="i"
      />
    </q-btn-stub>
  </q-item-section-stub>
   
  <player-name-stub
    playerid="12345"
  />
   
  <confirm-stub
    onno="[Function]"
    onyes="[Function]"
    text="Delete Player 1?"
  />
</div>
`;
